### ISSUE TITLE
[ASEditableTextNode] Maximum number of lines to display

### ISSUE THREAD

**Issue by leotumwattana** [facebook/AsyncDisplayKit#2777]
## Request
I'm looking for a way to limit **ASEditableTextNode** so that it will size and display a maximum number of lines. 

## Behaviour
For example let's say `editableTextNode.maximumLinesToDisplay = 6`

This should then size the editableTextNode to a maximum of 6 lines. More specifically:

When there is only 3 lines worth of text, then it should size to three lines. However, when there are 8 lines worth of text, then it should size to six lines and become scrollable for users to see the 7th and 8th lines.

## Additional Notes:
Simply setting `editableTextNode.textView.textContainer.maximumNumberOfLines = 6` will not work in this case, as although this will limit the textView to render only up to six lines, it will also not allow users to continue typing and see beyond those six lines.

Calculating using `font.lineHeight` may not work either, as that will unlikely take into account paragraph settings, etc.

What I did with plain UIKit is to use a TextKit stack to calculate the height of the UITextView using `layoutManager.lineFragmentRect`

Thanks in advance!

**Comment by EviluS** [comment]
@leotumwattana In one of our projects I had to implement a similar feature. The behavior was a single line of editable text node and when you start typing it will grow up to three lines. When it exceeds the max number of lines it will stop growing and start scrolling. 

The implementation is:
- set a preferred frame size for the editable text node (defining the max height)
- adopt the `ASEditableTextNodeDelegate`
- implement the `editableTextNodeDidUpdateText:` method
- create a variable in which you will save the calculated content size of the editable text node
- in each call of the `editableTextNodeDidUpdateText:` check whether the text has grown and relayout the text node.

Basic example:
- Lets say that we need maximum of 3 lines and assume that the point size of the font is 20px so our max frame is 60px height.
- If we have 1 line we will layout our node to fit.
- if we have more than 2 lines we will limit our node to the max height so that it can enable its scrolling.

Sorry in advance if my implementation is not explained clearly. There are probably better implementations but this one works for men. If you need any additional help you can contact me in the ASDK Slack I am `gpetrov` there.

**Comment by leotumwattana** [comment]
@EviluS Thanks. I'm thinking of making it a little more seamless, since this really does sound like something more than I alone will use. As @appleguy mentioned in his talks, `ASKD` is now going beyond just providing performance, but also convenience for framework users.

Looking through ASEditableTextNode.mm, I'm thinking of extending `ASTextKitComponents` with something like:

```
- (CGSize)sizeForConstrainedWidth:(CGFloat)constrainedWidth
                  forMaxNumberOfLines:(NSInteger)numberOfLines
```

Using `layoutManager.lineFragmentRect` to loop through and find the line fragmentRect that corresponds to the maxLine. That should give me a more accurate height than using a fixed font's point height. Especially since I'm expecting mixed fonts, use of Emojis, Chinese Characters, paragraph styles, etc.

Anyway, with that in place, I can modify ASEditableTextNode's `- (CGSize)calculateSizeThatFits:(CGSize)constrainedSize` to calculate the height using the existing method if `maximumLinesToDisplay = 0` or the above if `maximumLinesToDisplay != 0`.

Anyway, the only issue I have is that I'm absolutely not familiar with Objective-C. It generally makes my eyes glaze over 😝 . If this was Swift, then I'd be able to try this out right away. Haha. So now, I'm going to wait on my friend who's more Objective-C literate to help me out tomorrow (HKT). 😉 

@zammitjames @maicki: Please do let me know if you think the above sounds reasonable and whether you want me to change any naming, etc.

**Comment by Adlai-Holler** [comment]
Resolved in #2867

**Comment by skensell** [comment]
Hey guys, Awesome feature. Thanks @leotumwattana. There's a bug with it though: The way you are measuring text does not take into account the `textContainerInset` property.   For example, try `_textInputNode.textContainerInset = UIEdgeInsetsMake(4, 15, 4, 15);` and you will see that it doesn't expand even when there is one line of text + <30pts of new text. :(  Maybe I'll create a new issue...
